library/x509_create.c

```diff
@@ -125,81 +125,285 @@ static const x509_attr_descriptor_t *x509_attr_descr_from_name( const char *name
+        goto exit;
+    }
+    /*
+     * Check if it is a quoted string
+     */
+    else if( *( *cur ) == '\"' )
+    {
+        in_quot = 1;
+        ( *cur )++;
+    }
 
-            oid = attr_descr->oid;
-            s = c + 1;
-            in_tag = 0;
-            d = data;
+    while( *cur < end )
+    {
+        /*
+         * Check whether reached end of value to evaluate,
+         * whether it is a closing quotation, or a separator character
+         */
+        if( in_quot == 1 && *( *cur ) == '\"' )
+        {
+            /*
+             * Increase the current position to after the closing quotation
+             */
+            ( *cur )++;
+            /*
+             * Mark that quotation has ended.
+             */
+            in_quot = 0;
+            break;
         }
-
-        if( !in_tag && *c == '\\' && c != end )
+        else if( in_quot == 0 && *( *cur ) == ',')
         {
-            c++;
+            break;
+        }
 
-            /* Check for valid escaped characters */
-            if( c == end || *c != ',' )
+        if( *cur != end && * ( *cur ) == '\\' )
+        {
+            ( *cur )++;
+            ret = x509_dn_check_special_chars( cur, end );
+            if( ret != 0 )
             {
                 ret = MBEDTLS_ERR_X509_INVALID_NAME;
                 goto exit;
             }
         }
-        else if( !in_tag && ( *c == ',' || c == end ) )
+
+        /*
+         * Fill the current data pointer with the unescaped character.
+         */
+        *( d++ ) = *( *cur );
+        /*
+         * A whitespace can be part of the string,
+         * but it can also be a trailing whitespace, so don't set the
+         * value length until the character is  not a whitespace
+         */
+        if( *( *cur ) != ' ' )
+            value_len = d - data;
+
+        if( value_len == MBEDTLS_X509_MAX_DN_NAME_SIZE )
         {
-            mbedtls_asn1_named_data* cur =
-                mbedtls_asn1_store_named_data( head, oid, strlen( oid ),
-                                               (unsigned char *) data,
-                                               d - data );
+            ret = MBEDTLS_ERR_X509_INVALID_NAME;
+            goto exit;
+        }
 
-            if(cur == NULL )
-            {
-                return( MBEDTLS_ERR_X509_ALLOC_FAILED );
-            }
+        ( *cur )++;
+    }
 
-            // set tagType
-            cur->val.tag = attr_descr->default_tag;
+    if( value_len == 0 )
+    {
+        ret = MBEDTLS_ERR_X509_INVALID_NAME ;
+        goto exit;
+    }
 
-            while( c < end && *(c + 1) == ' ' )
-                c++;
+    /*
+     * Check if quotation has ended.
+     */
+    if( in_quot == 1 )
+    {
+        ret = MBEDTLS_ERR_X509_INVALID_NAME ;
+        goto exit;
+    }
 
-            s = c + 1;
-            in_tag = 1;
-        }
+    /*
+     * We have reached here in the following conditions:
+     * 1. *cur == end
+     * 2. in_quot == 1 && *( *cur ) == '"'
```

Comment:
Is this really the case? There is a check just before failing if `in_quot == 1`

Reply:
Yes, perhaps the description is not clear. 
When we are `in_quot` and reached a closing `"~ character, we set `in_quot` to false, and exit the parser loop. This is the second condition ( even if we already set `in_quot` to false.)
If at the end of the loop, `in_quot` is true, then we haven't got a closing `"` character, and thus fail

Reply:
Thanks for the explanation, I agree with your assessment. However, I also think that the comment is very misleading, because if I am not familiar with the code and read this comment, I will probably get confused with the condition immediately above.

To make things clearer, how about if instead of just writing the conditions we also write an english version of what we expect? For example, instead of just `*cur == end` we have `The end of the input buffer was reached (*cur == end)`.